## Final Project Submission

Please fill out:
* Student name: Benjamin McCarty
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: David Elliott
* Blog post URL:

---
> **[Return to Workflow](#return)**

---

# 📋 **To-Do's** 📋

---
**Priority**

- [ ] [WIP - Keep or delete](#wip) (will raise error!)
---
**Pending**
- [ ] Fix [guiding questions](#td-questions)
- [ ] Add links to each section addressing the rubric




# 📊 Phase 2 Project - DSC 060721 📊 

Author: Ben McCarty

Date: 7/12/21

Contact: bmccarty505@gmail.com

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Final-Project-Submission" data-toc-modified-id="Final-Project-Submission-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Final Project Submission</a></span></li></ul></li><li><span><a href="#📋-To-Do's-📋" data-toc-modified-id="📋-To-Do's-📋-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>📋 <strong>To-Do's</strong> 📋</a></span></li><li><span><a href="#📊-Phase-2-Project---DSC-060721-📊" data-toc-modified-id="📊-Phase-2-Project---DSC-060721-📊-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>📊 Phase 2 Project - DSC 060721 📊</a></span></li><li><span><a href="#📃-Rubric-" data-toc-modified-id="📃-Rubric--3"><span class="toc-item-num">3&nbsp;&nbsp;</span>📃 Rubric <a name="rubric"></a></a></span></li><li><span><a href="#🏡-(Re)Selling-Seattle:-Determining-House-Sell-Prices-🏡-" data-toc-modified-id="🏡-(Re)Selling-Seattle:-Determining-House-Sell-Prices-🏡--4"><span class="toc-item-num">4&nbsp;&nbsp;</span>🏡 <strong>(Re)Selling Seattle: Determining House Sell Prices</strong> 🏡 <a name="title"></a></a></span></li><li><span><a href="#❌-BEN---FIX-THIS!-❌" data-toc-modified-id="❌-BEN---FIX-THIS!-❌-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>❌ BEN - FIX THIS! ❌</a></span></li><li><span><a href="#📂-Setting-Up-the-Tools-and-Data" data-toc-modified-id="📂-Setting-Up-the-Tools-and-Data-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>📂 <strong>Setting Up the Tools and Data</strong></a></span><ul class="toc-item"><li><span><a href="#Imports" data-toc-modified-id="Imports-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#Importing-Personal-Functions" data-toc-modified-id="Importing-Personal-Functions-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Importing Personal Functions</a></span></li><li><span><a href="#Reading-Data" data-toc-modified-id="Reading-Data-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Reading Data</a></span></li></ul></li><li><span><a href="#🔮-Exploring-Fresh-Data" data-toc-modified-id="🔮-Exploring-Fresh-Data-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>🔮 <strong>Exploring Fresh Data</strong></a></span><ul class="toc-item"><li><span><a href="#Basic-Overviews" data-toc-modified-id="Basic-Overviews-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Basic Overviews</a></span></li><li><span><a href="#Data-Cleaning-and-Processing" data-toc-modified-id="Data-Cleaning-and-Processing-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Data Cleaning and Processing</a></span><ul class="toc-item"><li><span><a href="#Overview-Summary" data-toc-modified-id="Overview-Summary-7.2.1"><span class="toc-item-num">7.2.1&nbsp;&nbsp;</span>Overview Summary</a></span></li></ul></li><li><span><a href="#Exploring-Features" data-toc-modified-id="Exploring-Features-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Exploring Features<a name="features"></a></a></span></li></ul></li><li><span><a href="#🛠-Feature-Engineering" data-toc-modified-id="🛠-Feature-Engineering-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>🛠 <strong>Feature Engineering</strong></a></span><ul class="toc-item"><li><span><a href="#'yrs_old_sold'" data-toc-modified-id="'yrs_old_sold'-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span><code>'yrs_old_sold'</code></a></span><ul class="toc-item"><li><span><a href="#Determine-'year_sold'" data-toc-modified-id="Determine-'year_sold'-8.1.1"><span class="toc-item-num">8.1.1&nbsp;&nbsp;</span>Determine <code>'year_sold'</code></a></span></li><li><span><a href="#Calculate-'y_old_sold'" data-toc-modified-id="Calculate-'y_old_sold'-8.1.2"><span class="toc-item-num">8.1.2&nbsp;&nbsp;</span>Calculate <code>'y_old_sold'</code></a></span></li></ul></li><li><span><a href="#'was_renovated'" data-toc-modified-id="'was_renovated'-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span><code>'was_renovated'</code></a></span></li><li><span><a href="#&quot;yrs_since_reno&quot;" data-toc-modified-id="&quot;yrs_since_reno&quot;-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span><code>"yrs_since_reno"</code></a></span></li><li><span><a href="#&quot;has_bsmnt&quot;" data-toc-modified-id="&quot;has_bsmnt&quot;-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>"<code>has_bsmnt</code>"</a></span></li></ul></li><li><span><a href="#🔗-Correlations" data-toc-modified-id="🔗-Correlations-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>🔗 <strong>Correlations</strong></a></span><ul class="toc-item"><li><span><a href="#Determining-Correlations-with-Price" data-toc-modified-id="Determining-Correlations-with-Price-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Determining Correlations with Price</a></span></li><li><span><a href="#Determining-Multicollinearity" data-toc-modified-id="Determining-Multicollinearity-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Determining Multicollinearity</a></span></li></ul></li><li><span><a href="#🪓-Performing-Train/Test-Split" data-toc-modified-id="🪓-Performing-Train/Test-Split-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>🪓 <strong>Performing Train/Test Split</strong></a></span><ul class="toc-item"><li><span><a href="#Create-Train/Test-Split" data-toc-modified-id="Create-Train/Test-Split-10.1"><span class="toc-item-num">10.1&nbsp;&nbsp;</span>Create Train/Test Split</a></span></li></ul></li><li><span><a href="#📊-Creating-Baseline-Model" data-toc-modified-id="📊-Creating-Baseline-Model-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>📊 <strong>Creating Baseline Model</strong></a></span></li><li><span><a href="#🚿-Creating-Pipeline-for-Modeling" data-toc-modified-id="🚿-Creating-Pipeline-for-Modeling-12"><span class="toc-item-num">12&nbsp;&nbsp;</span>🚿 <strong>Creating Pipeline for Modeling</strong></a></span><ul class="toc-item"><li><span><a href="#Preprocessing-Data-Via-ColumnTransformer" data-toc-modified-id="Preprocessing-Data-Via-ColumnTransformer-12.1"><span class="toc-item-num">12.1&nbsp;&nbsp;</span>Preprocessing Data Via ColumnTransformer</a></span></li></ul></li><li><span><a href="#📊-Statsmodels-Non-Formula-OLS-Model" data-toc-modified-id="📊-Statsmodels-Non-Formula-OLS-Model-13"><span class="toc-item-num">13&nbsp;&nbsp;</span>📊 <strong>Statsmodels Non-Formula OLS Model</strong></a></span><ul class="toc-item"><li><span><a href="#Visualizing-zipcode-coefficients" data-toc-modified-id="Visualizing-zipcode-coefficients-13.1"><span class="toc-item-num">13.1&nbsp;&nbsp;</span>Visualizing zipcode coefficients</a></span></li><li><span><a href="#Visualizing-all-parameters-besides-zipcodes" data-toc-modified-id="Visualizing-all-parameters-besides-zipcodes-13.2"><span class="toc-item-num">13.2&nbsp;&nbsp;</span>Visualizing all parameters besides zipcodes</a></span></li></ul></li><li><span><a href="#🚿-Instantiating-Pipeline" data-toc-modified-id="🚿-Instantiating-Pipeline-14"><span class="toc-item-num">14&nbsp;&nbsp;</span>🚿 <strong>Instantiating Pipeline</strong></a></span><ul class="toc-item"><li><span><a href="#Non-Scaled-Pipeline" data-toc-modified-id="Non-Scaled-Pipeline-14.1"><span class="toc-item-num">14.1&nbsp;&nbsp;</span>Non-Scaled Pipeline</a></span></li></ul></li><li><span><a href="#✨-✅-Verifying-Lin-Reg-Assumptions-✨" data-toc-modified-id="✨-✅-Verifying-Lin-Reg-Assumptions-✨-15"><span class="toc-item-num">15&nbsp;&nbsp;</span>✨ ✅ Verifying Lin Reg Assumptions ✨</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Visualizing--Non-Scaled-Results" data-toc-modified-id="Visualizing--Non-Scaled-Results-15.0.1"><span class="toc-item-num">15.0.1&nbsp;&nbsp;</span>Visualizing  Non-Scaled Results</a></span></li></ul></li><li><span><a href="#Scaled-Pipeline" data-toc-modified-id="Scaled-Pipeline-15.1"><span class="toc-item-num">15.1&nbsp;&nbsp;</span>Scaled Pipeline</a></span></li></ul></li><li><span><a href="#✨-✅-Verifying-Lin-Reg-Assumptions-✨" data-toc-modified-id="✨-✅-Verifying-Lin-Reg-Assumptions-✨-16"><span class="toc-item-num">16&nbsp;&nbsp;</span>✨ ✅ Verifying Lin Reg Assumptions ✨</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Visualizing-Scaled-Results" data-toc-modified-id="Visualizing-Scaled-Results-16.0.1"><span class="toc-item-num">16.0.1&nbsp;&nbsp;</span>Visualizing Scaled Results</a></span></li></ul></li><li><span><a href="#Recommendations-Based-on-Modeling" data-toc-modified-id="Recommendations-Based-on-Modeling-16.1"><span class="toc-item-num">16.1&nbsp;&nbsp;</span>Recommendations Based on Modeling</a></span></li></ul></li><li><span><a href="#❌-RAISE-EXCEPTION-❌" data-toc-modified-id="❌-RAISE-EXCEPTION-❌-17"><span class="toc-item-num">17&nbsp;&nbsp;</span>❌ <strong>RAISE EXCEPTION</strong> ❌</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Interpretation-of-Model-Results" data-toc-modified-id="Interpretation-of-Model-Results-17.0.1"><span class="toc-item-num">17.0.1&nbsp;&nbsp;</span>Interpretation of Model Results</a></span></li></ul></li></ul></li><li><span><a href="#🥒-Pickling-the-Model-for-Reuse" data-toc-modified-id="🥒-Pickling-the-Model-for-Reuse-18"><span class="toc-item-num">18&nbsp;&nbsp;</span>🥒 <strong>Pickling the Model for Reuse</strong></a></span></li></ul></div>

# 📃 Rubric <a name="rubric"></a>

---
**Business Understanding:** [Business Understanding](#bizund)

- [ ] Notebook clearly explains the project's value for helping a specific stakeholder solve a real-world problem.


- [ ] Introduction explains the real-world problem the project aims to solve


- [ ] Introduction identifies stakeholders who could use the project and how they would use it


- [ ] Conclusion summarizes implications of the project for the real-world problem and stakeholders

---
**Data Understanding:**
- [ ] Notebook clearly describes the source and properties of the data to show how useful the data are for solving the problem of interest.


- [ ] Describe the data sources and explain why the data are suitable for the project


- [ ] Present the size of the dataset and descriptive statistics for all features used in the analysis


- [ ] Justify the inclusion of features based on their properties and relevance for the project


- [ ] Identify any limitations of the data that have implications for the project

---
**Data Preparation:** 
- [ ] Notebook shows how you prepare your data and explains why by including…


- [ ] Instructions or code needed to get and prepare the raw data for analysis


- [ ] Code comments and text to explain what your data preparation code does


- [ ] Valid justifications for why the steps you took are appropriate for the problem you are solving

---
**Modeling:**
- [ ] Notebook demonstrates an iterative approach to model building.


- [ ] Runs and interprets a simple, baseline model for comparison


- [ ] Introduces new models that improve on prior models and interprets their results


- [ ] Explicitly justifies model changes based on the results of prior models and the problem context


- [ ] Explicitly describes any improvements found from running new models

---
**Regression Results:**
- [ ] Notebook clearly presents two features with strong relationships to sale prices


- [ ] Explicitly identify the features and interpret their coefficients in plain language


- [ ] Explain the implications of results for the real-world problem the project aims to address


- [ ] Identify specific actions that stakeholders should take to leverage these results

---
**Visualization:**
- [ ] Notebook includes three relevant and polished visualizations of findings that…


- [ ] Help the project stakeholder understand the value or success of the project


- [ ] Have text and marks to aid reader interpretation, such as graph and axis titles, axis ticks and labels, or legend (varies by visualization type)


- [ ] Use color, size, and/or location to appropriately facilitate comparisons


- [ ] Are not cluttered, dense, or illegible

---
**Code Quality:**
- [ ] Code in notebook and related files meets professional standards (e.g. PEP 8)


- [ ] Code is easy to read, using comments, spacing, variable names, and function docstrings


- [ ] All code runs and no code or comments are included that are not needed for the project


- [ ] Code minimizes repetition, using loops, functions, and classes


- [ ] Code adapted from others is properly cited with author names and location of the cited material
---

# 🏡 **(Re)Selling Seattle: Determining House Sell Prices** 🏡 <a name="title"></a>

<a name='bizund'></a>

**<p style="text-align: center;">If you are considering selling your house, one of the highest priorities is getting the best price for it.</p>**

You want to make sure that the hedges are trimmed; that the basement isn't leaking; and that new coat of paint is covering up all of the "art" covering the walls from your kids.

*But what else can do you?* Would that extra bathroom under the stairs be a worthwhile addition? What about that addition you always talked about building?

---
**<p style="text-align: center;">When you are exploring the option of selling, you may ask such questions as:</p>**

>* How do the different aspects of your house impact the price?
>
>
>* What can you do to improve the house's value?


---
You may have some ideas already (and if you don't, watch a few episodes of "This Old House" and you will have *plenty* of inspiration).

But how do you *really* know what features are best? How can you be sure that your intuition or expectations are based on fact and not assumptions? **Turn to the data for answers!**

---
**<p style="text-align: center;">Revisiting King County</p>**

My [prior exploration](https://github.com/BenJMcCarty/Phase_2_Project_Final) of the King County House Sales dataset yielded some recommendations, but I wanted to see if I can improve them. I am revisiting the data with new techniques and approaches to see how much I can improve the results.

---

# ❌ BEN - FIX THIS! ❌

**WHAT QUESTIONS ARE IMPORTANT FROM THE BUSINESS PERSPECTIVE? WHAT QUESTIONS DID I USE TO GUIDE MY EXPLORATION?** 

My project used data from house sales in King County, WA. to determine some of the top features determining the sell price of a home.

* **First, I explored the data using the Pandas package to review the data itself.** I reviewed the different features included in my dataset and their respective values; the descriptive statistics for the numerical data; and the overall size and shape of the data (how many rows and columns).

* **Then, I used Seaborn to plot the data and a fitted linear regression model for each feature against the sell price.** These visualizations helped differentiate which features to treat as *categorical* variables (e.g. features that would be one of a select number of options) versus *continuous* variables (which would have less of a rigid range of values).

* **In order to give more depth to my data, I engineered four new features based on the original features and data.** I determined the age of each house at sale; whether or not a home was renovated; how many years since a renovation; and whether or not a house had a basement.

* **I wanted to make sure all of the features and their data were relevant to apply to my future model.** I performed correlational comparisons to determine which features were too closely related (indicating multicollinearity), which would affect my modeling process later on.

* **To test my assumptions that certain features would have a statistically significant impact on the price, I performed a few statistical tests.** I performed an ANOVA test on our "condition" feature (comparing each ranking of condition against the others) and I ran a two-sample t-test on the "waterfront" and "was_renovated" features as well.

* **LReg for Inferences**

# 📂 **Setting Up the Tools and Data**

## Imports

In [75]:
# Data Handling
import pandas as pd
import numpy as np
from scipy import stats

# Visualizations
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib as mpl

# Modeling - Statsmodels
import statsmodels.formula.api as smf
import statsmodels.api as sms

# Modeling - SKLearn
from sklearn import metrics
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.feature_selection import RFECV
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.dummy import DummyRegressor
from sklearn import set_config

# Settings
%matplotlib inline
plt.style.use('seaborn-talk')
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')
set_config(display='diagram')

## Importing Personal Functions

I created several functions specific to this project and saved them in an external file. The file is available from the main directory.

In [76]:
%load_ext autoreload
%autoreload 2

from p2pr_functions.functions import *
import p2pr_functions.functions as pf

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [77]:
## Uncomment cell to review contents of module
# help(pf)

## Reading Data

In [78]:
## Reading in data as DataFrame
local_data = 'bakeoff_data/Xtrain.csv'

df = pd.read_csv(local_data, index_col=0)

# 🔮 **Exploring Fresh Data**

## Basic Overviews

The goal for this section is to get a broad idea of the data before I start any cleaning or feature engineering.

In [79]:
df.head()

bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
date                                                                         
3/4/2015           3       2.50         1880      4499    2.00        0.00   
10/7/2014          3       2.50         2020      6564    1.00        0.00   
1/16/2015          5       4.00         4720    493534    2.00        0.00   
3/30/2015          2       2.00         1430      3880    1.00        0.00   
10/14/2014         3       2.25         2270     32112    1.00        0.00   

            view  condition  grade  sqft_above sqft_basement  yr_built  \
date                                                                     
3/4/2015    0.00          3      8        1880           0.0      1993   
10/7/2014   0.00          3      7        1310         710.0      1994   
1/16/2015   0.00          5      9        3960         760.0      1975   
3/30/2015   0.00          4      7        1430           0.0      1949   
10/14/2014  0.00          4      8        1740         530.0      1980   

            yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15  
date                                                                        
3/4/2015            0.00    98029 47.57 -122.00           2130        5114  
10/7/2014           0.00    98042 47.35 -122.16           1710        5151  
1/16/2015           0.00    98027 47.45 -122.01           2160      219542  
3/30/2015           0.00    98117 47.68 -122.39           1430        3880  
10/14/2014          0.00    98042 47.35 -122.09           2310       41606

In [80]:
## Review summary of dataframe details
report_df(df)

(16197, 18)


datatypes  num_unique  null_sum  null_pct     count      mean  \
bedrooms          int64          11         0      0.00 16,197.00      3.37   
bathrooms       float64          26         0      0.00 16,197.00      2.12   
sqft_living       int64         896         0      0.00 16,197.00  2,083.69   
sqft_lot          int64        8015         0      0.00 16,197.00 15,071.89   
floors          float64           6         0      0.00 16,197.00      1.49   
waterfront      float64           2      1756      0.11 14,441.00      0.01   
view            float64           5        49      0.00 16,148.00      0.23   
condition         int64           5         0      0.00 16,197.00      3.41   
grade             int64          11         0      0.00 16,197.00      7.66   
sqft_above        int64         814         0      0.00 16,197.00  1,790.47   
sqft_basement    object         280         0      0.00       nan       nan   
yr_built          int64         116         0      0.00 16,197.00  1,971.02   
yr_renovated    float64          66      2879      0.18 13,318.00     81.99   
zipcode           int64          70         0      0.00 16,197.00 98,078.10   
lat             float64        4783         0      0.00 16,197.00     47.56   
long            float64         720         0      0.00 16,197.00   -122.21   
sqft_living15     int64         691         0      0.00 16,197.00  1,987.81   
sqft_lot15        int64        7230         0      0.00 16,197.00 12,784.07   

                    std       min       25%       50%       75%          max  
bedrooms           0.91      1.00      3.00      3.00      4.00        11.00  
bathrooms          0.77      0.50      1.75      2.25      2.50         8.00  
sqft_living      918.21    370.00  1,430.00  1,912.00  2,560.00    13,540.00  
sqft_lot      40,775.85    520.00  5,058.00  7,620.00 10,720.00 1,651,359.00  
floors             0.54      1.00      1.00      1.50      2.00         3.50  
waterfront         0.09      0.00      0.00      0.00      0.00         1.00  
view               0.77      0.00      0.00      0.00      0.00         4.00  
condition          0.65      1.00      3.00      3.00      4.00         5.00  
grade              1.17      3.00      7.00      7.00      8.00        13.00  
sqft_above       827.60    370.00  1,200.00  1,560.00  2,220.00     9,410.00  
sqft_basement       nan       nan       nan       nan       nan          nan  
yr_built          29.33  1,900.00  1,952.00  1,975.00  1,997.00     2,015.00  
yr_renovated     396.21      0.00      0.00      0.00      0.00     2,015.00  
zipcode           53.49 98,001.00 98,033.00 98,065.00 98,117.00    98,199.00  
lat                0.14     47.16     47.47     47.57     47.68        47.78  
long               0.14   -122.52   -122.33   -122.23   -122.12      -121.31  
sqft_living15    685.19    399.00  1,490.00  1,840.00  2,360.00     6,210.00  
sqft_lot15    26,833.38    651.00  5,100.00  7,620.00 10,086.00   871,200.00

## Data Cleaning and Processing

---
**DF Columns to Convert**

* 'date' to datetime; continuous variable


Potential Changes:

>* 'waterfront' to category; nominal/binary variable
>* 'view' to category; ordinal
>* 'condition' to category; ordinal variable
>* 'grade' to category; ordinal variable
>* 'zipcode' to category; nominal variable


---
**Null Values to Process**

Columns with null values:
* 'yr_renovated'
* 'waterfront
* 'view'

Options:
>* Fill with mean/median/mode
>* Fill with zeroes
>
>
>* **PIPELINES!!!**
>    * SimpleImputer
>    * IterativeImputer *(Experimental)*
>    * GridSearchCV for hyperparameter tuning

---

In [81]:
## Converting 'date' column to datetime

df['date'] = pd.to_datetime(df['date'])
df['date']

KeyError: 'date'

In [ ]:
## Converting 'sqft_basement' to numeric and filling any null values with zero

df['sqft_basement'] = pd.to_numeric(df['sqft_basement'], errors='coerce')
df['sqft_basement']

In [ ]:
##Filling null values with the most frequent value

for col in df:
    if df[col].isna().sum() > 0:
        df[col].fillna(df[col].mode()[0], inplace=True)

In [ ]:
## Reviewing dataframe post-fill

report_df(df)

### Overview Summary

The dataset contains 20 columns of data, most of which seem useful for evaluations and modeling.

I manually fixed null values at this stage - I will start the pipeline later on.

After processing, I do not have any null values and all of the datatypes are correct (I can convert some columns into the "category" datatype for analysis later, if needed).

## Exploring Features<a name='features'></a>

Now I will be comparing each individual feature against "price" to determine if a feature is continuous or categorical.

I will check the regressions and histograms for each feature; if a regression plot shows a flatter regression line or if the histogram is not a normal distribution, I will treat that feature as a categorical feature.

As part of the function to create the plots, I include my function to ignore any outliers for the purpose of the visualizations. I will *not* change the data itself, though.

***Tip:*** *Links are included below for quick navigation to the different features.*

<a name='features'></a>

| Feature Names | Feature Names  |
| ------------ | ------------- |
| [Bedrooms](#bedrooms)     | [Sqft_above](#above)    |
| [Bathrooms](#bathrooms)    | [Sqft_basement](#base) |
| [Sqft_living](#living)  | [Yr_built](#built)      |
| [Sqft_lot](#lot)     | [Yr_renovated](#reno)  |
| [Floors](#floors)       | [Zipcode](#zip)       |
| [Waterfront](#wf)   | [Lat](#lat)           |
| [View](#view)         | [Long](#long)          |
| [Condition](#cond)    | [Sqft_living15](#living15)    |
| [Grade](#grade)        | [Sqft_lot15](#lot15)    |

**Bedrooms** <a name="bedrooms"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(data=df,x='bedrooms', categorical=True, kde = False)

***
**Observations**
>* There is a slight linear regression between the number of bedrooms and the sell price, indicated by the positive slope of the regression line.
>
>
>* The histogram slightly resembles a normal distribution, although the first and last bins are lower than I would expect for a normal distribution.

***
**TO-DO**
>* Use `'bedrooms'` as a categorical variable.
***

**Bathrooms** <a name="bathrooms"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"bathrooms", categorical=True, kde=False)

***
**Observations**
>* There is a clear linear relationship between bathrooms and price.
>
>
>* The histogram/normal distribution isn't the strongest, particularly for the larger number of bathrooms.

***
**TO-DO**
>
>* Similar to bedrooms, the values for number of bathrooms are discrete and fall into specific values. This indicates that the feature is categorical.
*** 

**Sqft_living** <a name="living"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_living")

***
**Observations**

>* There is a clear linear relationship, although the distribution is slightly skewed to the right, indicating it's not a perfect normal distribution.
>
>
>* As the values are spread across the whole range, this feature is clearly a continuous variable.

***

**To-Do**

>* Use as continuous variable
***


**Sqft_lot** <a name="lot"></a>

In [ ]:
show_cleaned_vis(df,"sqft_lot")

***
**Observations**

>* Weak linear relationship with price; distribution skewed right.
***

**To-Do**

>* Treat as continuous variable as the values are spread out across the whole range.

***

**Floors** <a name="floors"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"floors", kde=False)

***
**Observations**

>* Linear relationship, but clearly categorical as the values fall into specific categories ranging from 1-3.5
>
>* Most homes had one or two floors.

***
**To-Do**

>* Treat as categorical variable.
 
***

**Waterfront** <a name="wf"></a> [🔝](#features)

In [ ]:
feature_vis(df,"waterfront", categorical=True, kde=False)

***
**Observations**

>* Graphs are misleading - clearly binary variable with a slight linear relationship.
>
>
>* Most properties are non-waterfront.
>
>
>* Clear linear relationship, but not normally distributed

***
**To-Do**

>* Treat as categorical - only two options (0 or 1).
 
***

**View** <a name="view"></a> [🔝](#features)

In [ ]:
feature_vis(df,"view", categorical=True, kde=False)

***
**Observations**

>* Slight linear relationship - higher "view" rating, higher price
>
>
>* Most homes have '0' view
>
>
>* A few extreme outliers in pricing (for values 2, 3, and 4s)

***
**To-Do**

>* Treat as categorical
 
***

**Condition** <a name="cond"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"condition", categorical=True, kde= False)

***
**Observations**

>* Largest number of homes sold were in condition 3, followed by condition 4.
>
>
>* Very few homes sold with conditions of 1s or 2s.
>
>
>* Price outliers in 4.0 area, some slight outliers in 2 and 3
>
>
>* No linear relationship, not normal distribution.

***
**TO-DO**

>* Treat as categorical
 
***

**Grade** <a name="grade"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"grade", kde=False)

**Observations**

>* Clear linear trend - as grade increases, so does price
>
>
>* Grades range from 6-9
>
>
>* Price outliers for grades 7 and 8

---
**To-Do**

>* Treat as categorical variable


**Sqft_above** <a name="above"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_above")

---

**Observations**

>* Outliers impacting accuracy of linear regression, but still seems relatively linear.
>
>
>* Distribution skewed right, but otherwise normal
---
**TO-DO**

>* Treat as continuous
---

**Sqft_basement** <a name="base"></a>[🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_basement")

---
**Observations**

>* There is a very large number of 0 ft$^2$ basements, indicating that these homes do not have a basement
>
>
>* There are scattered outliers in pricing
>
>
>* The regression is poor due to the large number of 0 values

---
**To-Do**

>* Treat as a continuous variable
>
>
>* Can create new categorical, binary variable to indicate whether or not a house has a basement to measure the impact on price.
---

**Yr_built** <a name="built"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"yr_built")

---
**Observations**

>* No linear trend from this regression plot
>
>* Seems like may houses built between 1940 - 1970, then major boom in early 2000s.
>
>* Some significant outliers in price, but not enough to  impact the regression
---

**To-Do**

>* Compare year built to year sold - how old was the house at sale?
>
>
>* Potentially discretize for analysis.
---

**Yr_renovated** <a name="reno"></a> [🔝](#features)

In [ ]:
feature_vis(df,'yr_renovated', kde=False)

---
**Observations**

>* Number of houses not renovated significantly outweighs the number renovated
>
>
>* Graphs indicate that an engineered feature for whether or not the house was renovated would be worthwhile.

---
**To-Do**

>* Create new categorical feature - renovated or not
---

**Zipcode** <a name="zip"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"zipcode", categorical=True, kde=False)

---
**Observations**

>* No linear relationship; not normally distributed
---
**TO-DO**

>* Treat as categorical variable
>
>* Perform OneHotEncoding to create new features for modeling.
---

**Lat** <a name="lat"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"lat")

---
**Observations**

>* Very weak linear relationship and not normally distributed.
>
>
>* Outliers just slightly past 47.6 degrees latitude

---
**To-Do**

>* Treat as categorical for modeling.
---

**Long** <a name="long"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"long")

---
**Observations**

>* No linear relationship and not normally distributed.
>
>
>* Outliers around -122.32ish and -122.25-ish degrees

---
**To-Do**

>* Treat as continuous if used for modeling.
---

**Sqft_living15** <a name="living15"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_living15")

---
**Observations**

>* Most properties sold between 1500/2000 ft$^2$ living space
>
>
>* Price follows linear trend and is normally distributed, although skewed right.
>
>* Outliers between 2500-3500 ft$^2$ 
---
**TO-DO**

>* Treat as continuous variable
---

**Sqft_lot15** <a name="lot15"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_lot15")

---
**Observations**

>* No clear linear trend
>* Significant outlier at approx. 9000 ft$^2$

---
**TO-DO**
>* Treat as continuous variable
---

# 🛠 **Feature Engineering**

## `'yrs_old_sold'`

I create this feature to differentiate between houses that were built recently versus older houses.

In order to determine this feature, I need to determine the year the house was sold first.

### Determine `'year_sold'`

In [ ]:
## Pull the year from the "date" column
df['year_sold'] = pd.DatetimeIndex(df['date']).year

## Review the values to ensure data integrity
df['year_sold'].value_counts()

### Calculate `'y_old_sold'`

In [ ]:
## Calculating the age of the house at the time of sale
df['y_old_sold'] = df['year_sold'] - df['yr_built']

## Minimum age is -1 due to a house being sold before it was finished being built
display(df['y_old_sold'].value_counts().sort_index(), df['y_old_sold'].describe())

>**Results:** Half of the homes sold fall between the ages of 18 and 63 years old, with the average age of 43 years. Our min and max ages indicate a house was sold before it was built, while another house was 115 years old.

## `'was_renovated'`

What impact would a renovation have on the price?

In [ ]:
## Using the year that the home was renovated to deterine whether or not the home was renovated
reno_y_n = np.where(df['yr_renovated']>0, 1, 0 )
df = df.assign(was_renovated = reno_y_n)
df.head(5)

In [ ]:
## Checking values
df['was_renovated'].value_counts()

>**Results:** Most of the houses were not renovated at the time of sale. This feature may not have a significant impact on determining the sale price due to the small number of renovated houses.

## `"yrs_since_reno"`

If a house was renovated, how long ago was the renovation? Would more newly-renovated houses increase price?

In [ ]:
## Ensuring there are no null values in the new feature and replacing any with zeroes

df['yrs_since_reno'] = np.where((df['was_renovated']==1),
                                (df['year_sold'] - df['yr_renovated']), 0)

display(df['yrs_since_reno'].describe(),df['yrs_since_reno'].value_counts(ascending=False))

In [ ]:
## Checking for which properties show a renovation year post-sale

df[df['year_sold'] < df['yr_renovated']]

In [ ]:
## Reviewing stats for those houses that were renovated

yrs_and_reno = df[df['yrs_since_reno'] > 0]
yrs_and_reno['yrs_since_reno'].describe()

>**Results:** As expected, most of the houses were not renovated (indicated by the number of years being zero). 
>
>
>Of those renovated, half of the homes had between 9 and 28 years between the years sold and renovated, with an average ages of roughly 20 years old.
>
>
>Interestingly, this dataset also shows *six records for which the renovation is shown to be post-sale.* While this may be an error, I am leaving it in the dataset since I cannot confidently rule it out. 
>
>
>If this feature does not show statistical significance, I may drops these values and reevaluate the significance.

## "`has_bsmnt`"

I noticed that there were fewer houses with a value for "sqft_basement" during my data exploration. I am curious if the presence or absence of a basement would have any impact.

In [ ]:
## Determining whether or not a house has a basement based on the square footage
df['has_bsmnt'] = np.where(df['sqft_basement'] > 0, 1, 0)

# Reviewing the results
display(df['has_bsmnt'].describe(), df['has_bsmnt'].value_counts())

# 🔗 **Correlations**

## Determining Correlations with Price

In [ ]:
## Determining top five features most strongly correlated with price
## (Considering absolute values for magnitude, not direction)

df_corr = np.abs(df.drop(['price', 'lat','long', 'yr_renovated'], axis=1).corrwith(df['price']).sort_values(ascending=False))
display(df_corr[0:5])

## Determining Multicollinearity

In [ ]:
## Creating correlation matrix w/ heatmap for intensity of values
corr_val(df.drop(['price'],axis=1), figsize=(20,20));

In [ ]:
## Correlation results ignoring duplicate values 
## Except for "was_renovated", which shows a perfect relationship to itself

df_corr_results = np.abs(df.drop(['price','yr_renovated'],axis=1).corr().unstack().sort_values(ascending=False).drop_duplicates())

## Show strongest correlations
df_corr_results[1:11]

**Results:** From the correlation matrix, I see that we have seven pairs of features that show a correlation over .75.

I will drop 'sqft_living', 'was_renovated', and 'has_basement' due to their high correlations.

In [ ]:
# Dropping columns to address multicollinearity (over .75)
df_dropped = df.drop(['was_renovated', 'sqft_living','has_bsmnt'], axis=1)
# df_dropped

In [ ]:
# Rerunning model
corr_val(df_dropped.drop('price',axis=1), figsize=(20,20));

In [ ]:
# Correlation results ignoring (most) duplicate values
df_corr_results = df_dropped.drop('price',axis=1).corr().unstack().sort_values(ascending=False).drop_duplicates()

## Show strongest correlations
print(df_corr_results[1:11],'\n\n',df_corr_results[-11:-1])

# 🪓 **Performing Train/Test Split**

In [ ]:
## Creating features matrix minus target variable
X = df.drop('price', axis = 1).copy()

In [ ]:
## Creating the y values by setting them equal to the 'price' values from the dataframe
y = df['price'].copy()

In [ ]:
## Verifying the two groups are of equal length
X.shape[0] == y.shape[0]

## Create Train/Test Split

Now I will split the data into the train/test groups. Then, I will run the first linear regression on the "train" data, then another regression on the "test" data.

In [ ]:
## Establishing the train and test data before doing anything else

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,
                                                    random_state=505)

# 📊 **Creating Baseline Model**

In [ ]:
baseline = DummyRegressor()

baseline.fit(X, y)
print(f'The baseline r^2 score is: {baseline.score(X, y):.2f}')

<!-- Do not delete this cell! It has anchors for other links -->

<a name="eod"></a>
<a name="pipeline_2"></a>

# 🚿 **Creating Pipeline for Modeling**

**For original code:** [Click here](#pipeline_1)

## Preprocessing Data Via ColumnTransformer

In [ ]:
## Specifying features for each transformer (below)

imp_zero_cols = ['sqft_basement']
imp_mode_cols = ['view','waterfront','yr_renovated']
ohe_cols = ['condition','grade', 'zipcode']

xf_cols = [*imp_zero_cols, *imp_mode_cols, *ohe_cols]

remaining_cols = [i for i in X.drop('date', axis=1).columns if i not in xf_cols]

imp_zero_cols.extend(remaining_cols)

In [ ]:
## Creating ColumnTransformer and sub-transformers

#----------------------------------------------------------------------------#

# ## Removed this code due to errors encountered b/t train, test categoricals
# ohe_cols = ['condition','grade', 'zipcode']
# categorical_transformer = OneHotEncoder(drop='first', sparse=False)

# ## A backup transformer to future-proof against null values in unseen data
# ## Removed - throwing errors when fitting
# remaining_transformer = SimpleImputer(strategy='most_frequent')
# ,('remaining', remaining_transformer, remaining_cols)

#----------------------------------------------------------------------------#

# Adding zeroes for missing values in sqft_basement
zero_transformer = SimpleImputer(strategy='constant', fill_value=0)

## Adding the mode value to view, waterfront, and yr_renovated missing values 
mode_transformer = SimpleImputer(strategy='most_frequent')

## Encoding categoricals - handling errors to prevent issues w/ test set
categorical_transformer = OneHotEncoder(handle_unknown='ignore', sparse=False)

## Instantiating the ColumnTransformer and including all transformers
preprocessor = ColumnTransformer(
    transformers=[
        ('zero', zero_transformer, imp_zero_cols),
        ('mode', mode_transformer, imp_mode_cols),
        ('cats', categorical_transformer, ohe_cols)],)

preprocessor

In [ ]:
## Fit and transform the data via the ColumnTransformer
X_train_tf = preprocessor.fit_transform(X_train)

# 📊 **Statsmodels Non-Formula OLS Model**

>I create a Statsmodels version of the regression model to compare against my later SKLearn models. I use this comparison to ensure that the SKLearn pipeline below creates the same predictions for the same transformed data

In [ ]:
## Getting feature names from OHE
ohe_cat_names = preprocessor.named_transformers_['cats'].get_feature_names(ohe_cols)

## Generating list for column index
final_cols = [*imp_zero_cols, *imp_mode_cols, *ohe_cat_names]

## Instantiating a DF storing the resulting transformed training data
X_train_tf_df = pd.DataFrame(X_train_tf, columns=final_cols, index=X_train.index)

## Transforming the test set
X_test_tf = preprocessor.transform(X_test)

## Creating DF for test data
X_test_tf_df = pd.DataFrame(X_test_tf, columns=final_cols, index=X_test.index)

display(X_train_tf_df.head(5),X_test_tf_df.head(5))

In [ ]:
## Running a Statsmodels OLS model for verification
sm_reg = sms.OLS(y_train, sms.add_constant(X_train_tf_df)).fit()

diagnose_model(sm_reg)

In [ ]:
## Pulling SM-OLS coefficients
model_params = sm_reg.params
model_params

## Visualizing zipcode coefficients

In [ ]:
## Creating list of zipcode parameters for filtering
zipcode_cols = [x for x in model_params.index if x.startswith('zipcode')]

In [ ]:
## Creating visualization for zipcodes
only_zips = model_params[zipcode_cols]

plot_coefs(data = only_zips, kind = "bar", figsize=(25,10), x_label = 'Zipcodes',
           y_label = 'Coefficient ($)', title= 'Zip Code Effect on Price');

## Visualizing all parameters besides zipcodes

In [ ]:
## Plotting all features besides zip codes
no_zips = model_params.drop(zipcode_cols)
no_zips = no_zips.drop('const')

plot_coefs(data = no_zips, kind = "barh", x_label = 'Features Other Than Zip Code',
           y_label = 'Coefficient ($)', title= 'House Feature Effect on Price',
          figsize = (10,15));

# 🚿 **Instantiating Pipeline**

## Non-Scaled Pipeline

In [ ]:
## Creating full pipeline with ColumnTransformer and LinearRegression
reg_pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', LinearRegression())])

In [ ]:
## Fitting pipeline to the training data
reg_pipe.fit(X_train, y_train)

## Evaluating model performance
eval_perf_total(reg_pipe, X_train, y_train, X_test, y_test)

# ✨ ✅ Verifying Lin Reg Assumptions ✨

In [ ]:
y_test_pred = reg_pipe.predict(X_test)

In [ ]:
## Determining residuals
residuals = (y_test - y_test_pred)
residuals

In [ ]:
## Plotting to test for normality
sns.histplot(data=residuals);

In [ ]:
## Checking the homoscedasticity of the new model
sns.residplot(x=y_test, y=residuals, lowess=True, color="g");

### Visualizing  Non-Scaled Results

In [ ]:
reg_pipe_coefs = get_model_coefs(reg_pipe, final_cols)
only_zips = model_params[zipcode_cols]

reg_zips = reg_pipe_coefs[zipcode_cols]
reg_no_zips = reg_pipe_coefs.drop(zipcode_cols)
reg_no_zips = reg_no_zips.drop('intercept')

In [ ]:
plot_coefs(data = reg_zips, kind = "barh", figsize=(15,20), x_label = 'Zip Codes',
           y_label = 'Coefficient ($)', title= 'Effects of Zip Codes on Price')

plt.show()

plot_coefs(data = reg_no_zips, kind = "barh", figsize=(15,15), x_label = 'Features Other Than Zip Code',
           y_label = 'Coefficient ($)', title= 'Effects of House Features on Price');

## Scaled Pipeline

In [ ]:
## Creating full pipeline with ColumnTransformer, StandardScaler and LinearRegression
reg_pipe_scaled = Pipeline(steps=[('preprocessor', preprocessor),
                           ('scaler', StandardScaler()),
                      ('regressor', LinearRegression())])

In [ ]:
## Fittng pipeline to the training data
reg_pipe_scaled.fit(X_train, y_train)

## Ensuring same results as prior pipeline
eval_perf_total(reg_pipe, X_train, y_train, X_test, y_test)

# ✨ ✅ Verifying Lin Reg Assumptions ✨

In [ ]:
y_test_pred = reg_pipe_scaled.predict(X_test)

In [ ]:
## Determining residuals
residuals = (y_test - y_test_pred)
residuals

In [ ]:
## Plotting to test for normality
sns.histplot(data=residuals);

In [ ]:
## Checking the homoscedasticity of the new model
sns.residplot(x=y_test, y=residuals, lowess=True, color="g");

### Visualizing Scaled Results

In [ ]:
## Pulling model coefficients for visualizations

reg_pipe_scaled_coefs = get_model_coefs(reg_pipe_scaled, final_cols)

reg_scaled_zips = reg_pipe_scaled_coefs[zipcode_cols]

reg_scaled_no_zips = reg_pipe_scaled_coefs.drop(zipcode_cols)
reg_scaled_no_zips = reg_scaled_no_zips.drop('intercept')

In [ ]:
## Plotting coefficients

plot_coefs(data = reg_scaled_zips, kind = "barh", figsize=(20,25), x_label = 'Zip Codes',
           y_label = 'Zip Codes', title= 'Zip Code Effect on Price ($)')

plt.show()

plot_coefs(data = reg_scaled_no_zips, kind = "barh", figsize=(20,20), x_label = 'Features Other Than Zip Code',
           y_label = 'Features', title= 'Effects of House Features on Price ($)');

## Recommendations Based on Modeling

If a homeowner is interested in renovating their house, I would recommend the following improvements based on the high coefficients from my model.

* Add bathrooms
* Also consider adding bedrooms
* Use high-quality materials in renovations

While it is outside of the homeowner's ability to change, the zip code for a house's location also has a large impact on the price. A prospective buyer should keep the zip code of the area in mind to help determine the prices of the homes there.

In [ ]:
## Visualizing the impact of number of bathrooms on price
sns.regplot(data=df_clean, x="bathrooms", y='price')
plt.suptitle('Impact of Number of Bathrooms on Price')
plt.xlabel('Number of Bathrooms')
plt.ylabel('Price ($)');

In [ ]:
## Visualizing impact of number of bedrooms on price
sns.regplot(data=df_clean, x="bedrooms", y='price')
plt.suptitle('Impact of Number of Bedrooms on Price')
plt.xlabel('Number of Bedrooms')
plt.ylabel('Price ($)');

In [ ]:
## Visualizing impact of the grade of construction materials on price
sns.regplot(data=df_clean, x="grade", y='price')
plt.suptitle('Impact of Grade of Materials on Price')
plt.xlabel('Grade of Materials')
plt.ylabel('Price ($)');

In [ ]:
## Representing the diversity of prices based on zip codes

fg = sns.catplot(data=df_clean, x="zipcode", y='price', aspect=2.75, height=5)
fg.ax.set_xticklabels(fg.ax.get_xticklabels(), rotation=45, ha='right')
plt.suptitle('Impact of Zip Code on Price')
plt.xlabel('Zip Code')
plt.ylabel('Price ($)');

# ❌ **RAISE EXCEPTION** ❌

### Interpretation of Model Results

The model does a great job of explaining the variance between our selected features, with an r-squared value of .79/1.

The RMSE for both the training and test sets are below our target of 200,000! Additionally, the difference between the two is within an acceptable range.

The assumption tests confirm the assumptions that our residuals would be normally distributed and homoscedastic.

# 🥒 **Pickling the Model for Reuse**

In [ ]:
# ## Setting coefficients and target
# X = df[['was_renovated','waterfront', 'zipcode',
#             'yrs_since_reno', 'bedrooms', 'bathrooms', 'condition','grade']].copy()
# y = df['price'].copy()

In [ ]:
# ## Create list of categorical variables
# cat_feats = ['condition','grade', 'zipcode']

# ## Create the OHE without "drop='first" as it would throw an error in this case
# ohe = OneHotEncoder(handle_unknown = 'ignore', sparse=False)

# ## Using OHE on our categorical variables for training (NOT testing)
# final_ohe = ohe.fit_transform(X[cat_feats])

In [ ]:
# ## Getting feature names from our list of categories
# feat_col_name = ohe.get_feature_names(cat_feats)

In [ ]:
# ## Creating DF of final results
# penultimate_ohe_df = pd.DataFrame(final_ohe, columns=feat_col_name, index=X.index)

In [ ]:
# ## Updating final DF
# final_ohe_df = pd.concat([X.drop(cat_feats, axis=1), penultimate_ohe_df], axis=1)

In [ ]:
# ## Instantiating and fitting the regression
# lr_final = LinearRegression()

# lr_final = lr_final.fit(final_ohe_df, y)

In [ ]:
# # Generating predictions
# final_pred = lr_final.predict(final_ohe_df)

In [ ]:
# ## Confirming predictions work
# display(final_pred)

In [ ]:
# # Saving model to .pickle file
# import pickle

# pickle_out = open("final_model.pickle","wb")
# pickle.dump(lr_final, pickle_out)
# pickle_out.close()